packages/modelence/src/auth/verifyEmail.ts

```diff
@@ -0,0 +1,68 @@
+import { z } from 'zod';
+
+import { Args, Context } from '../methods/types';
+import { usersCollection, tokensCollection } from './db';
+import { setSessionUser } from './session';
+
+export async function handleVerifyEmail(args: Args, { session }: Context) {
+  if (!session) {
+    throw new Error('Session is not initialized');
+  }
+
+  const token = z.string().parse(args.token);
+
+  // Find token in database
+  const tokenDoc = await tokensCollection.findOne({
+    token,
+    type: 'emailVerification',
+    expiresAt: { $gt: new Date() }
+  });
+
+  if (!tokenDoc) {
+    throw new Error('Invalid or expired verification token');
+  }
+
+  // Find user by token's userId
+  const userDoc = await usersCollection.findOne({ _id: tokenDoc.userId });
+
+  if (!userDoc) {
+    throw new Error('User not found');
+  }
+
+  const email = tokenDoc.email;
+
+  if (!email) {
+    throw new Error('Email not found in token');
+  }
+
+  // Find the specific email address to verify
+  const emailIndex = userDoc.emails?.findIndex(e => e.address === email);
+  
+  if (!userDoc.emails || emailIndex === undefined || emailIndex === -1) {
+    throw new Error('Email address not found for this user');
+  }
+
+  if (userDoc.emails[emailIndex].verified) {
+    throw new Error('Email is already verified');
+  }
+
+  // Mark the specific email as verified
+  await usersCollection.updateOne(
+    { _id: tokenDoc.userId },
+    { $set: { [`emails.${emailIndex}.verified`]: true } }
+  );
```

Comment:
This part looks very flaky, and it is also very vulnerable to a race condition (if the emails list changes in between these operations everything will get mixed up).

MongoDB should already have a way to find & update a matching element in an array in a single command.
https://www.mongodb.com/community/forums/t/update-nested-sub-document-with-a-specific-condition/136373